odo/backends/sparksql.py

```diff
@@ -58,19 +91,55 @@ def rdd_to_sqlcontext(ctx, rdd, name=None, dshape=None, **kwargs):
         dshape = dshape.measure
     sql_schema = dshape_to_schema(dshape)
     sdf = ctx.applySchema(rdd, sql_schema)
-    ctx.registerRDDAsTable(sdf, name or next(_names))
+    if name is None:
+        name = next(_names)
+    ctx.registerDataFrameAsTable(sdf, name)
+    ctx.cacheTable(name)
     return sdf
 
 
 @discover.register(SQLContext)
 def discover_sqlcontext(ctx):
-    dshapes = valmap(discover, get_catalog(ctx))
-    return datashape.DataShape(datashape.Record(sorted(dshapes.items())))
+    table_names = list(map(str, ctx.tableNames()))
+    dshapes = sorted(zip(table_names,
+                         map(discover, map(ctx.table, table_names))))
+    return datashape.DataShape(datashape.Record(dshapes))
 
 
-@discover.register(SchemaRDD)
+@discover.register(SparkDataFrame)
 def discover_spark_data_frame(df):
-    return datashape.var * schema_to_dshape(df.schema())
+    return datashape.var * schema_to_dshape(df.schema)
+
+
+def chunk_file(filename, chunksize=1 << 30):
+    """Stream `filename` in chunks of size `chunksize`.
+
+    Parameters
+    ----------
+    filename : str
+        File to chunk
+    chunksize : int, optional, defaults to 1GB
+        Number of bytes to hold in memory at a single time
+    """
+    with open(filename, mode='rb') as f:
+        for chunk in iter(partial(f.read, chunksize), ''):
+            yield chunk
+
+
+@append.register((JSONLines, JSON), SparkDataFrame)
+def spark_df_to_jsonlines(js, df, pattern='part-*', **kwargs):
+    tmpd = tempfile.mkdtemp()
+    try:
+        df.save(tmpd, source='org.apache.spark.sql.json', mode='overwrite')
```

Comment:
It writes JSON as a directory of newline delimited JSON, one for each partition.

More info here: http://stackoverflow.com/a/23902886/564538